include/pmacc/particles/memory/dataTypes/SuperCell.hpp

```diff
@@ -22,70 +22,93 @@
 #pragma once
 
 #include "pmacc/types.hpp"
+#include "pmacc/math/vector/compile-time/Vector.hpp"
 
-namespace pmacc
-{
 
-template <class TYPE>
-class SuperCell
+namespace pmacc
 {
-public:
-
-    HDINLINE SuperCell() :
-    firstFramePtr(nullptr),
-    lastFramePtr(nullptr),
-    mustShiftVal(false),
-    sizeLastFrame(0)
-    {
-    }
-
-    HDINLINE TYPE* FirstFramePtr()
-    {
-        return firstFramePtr;
-    }
-
-    HDINLINE TYPE* LastFramePtr()
-    {
-        return lastFramePtr;
-    }
-
-    HDINLINE const TYPE* FirstFramePtr() const
-    {
-        return firstFramePtr;
-    }
 
-    HDINLINE const TYPE* LastFramePtr() const
+    template< class T_FrameType >
+    class SuperCell
     {
-        return lastFramePtr;
-    }
-
-    HDINLINE bool mustShift()
-    {
-        return mustShiftVal;
-    }
-
-    HDINLINE void setMustShift(bool value)
-    {
-        mustShiftVal = value;
-    }
-
-    HDINLINE lcellId_t getSizeLastFrame()
-    {
-        return sizeLastFrame;
-    }
-
-    HDINLINE void setSizeLastFrame(lcellId_t size)
-    {
-        sizeLastFrame = size;
-    }
-
-
-private:
-    PMACC_ALIGN(mustShiftVal, bool);
-    PMACC_ALIGN(sizeLastFrame, lcellId_t);
-public:
-    PMACC_ALIGN(firstFramePtr, TYPE*);
-    PMACC_ALIGN(lastFramePtr, TYPE*);
-};
-
-} //end namespace
+    public:
+
+        HDINLINE SuperCell() :
+            firstFramePtr( nullptr ),
+            lastFramePtr( nullptr ),
+            numParticles( 0 ),
+            mustShiftVal( false )
+        {
+        }
+
+        HDINLINE T_FrameType * FirstFramePtr()
+        {
+            return firstFramePtr;
+        }
+
+        HDINLINE T_FrameType * LastFramePtr()
+        {
+            return lastFramePtr;
+        }
+
+        HDINLINE T_FrameType const * FirstFramePtr() const
+        {
+            return firstFramePtr;
+        }
+
+        HDINLINE T_FrameType const*  LastFramePtr() const
+        {
+            return lastFramePtr;
+        }
+
+        HDINLINE bool mustShift() const
+        {
+            return mustShiftVal;
+        }
+
+        HDINLINE void setMustShift( bool const value )
+        {
+            mustShiftVal = value;
+        }
+
+        HDINLINE uint32_t getSizeLastFrame() const
+        {
+            constexpr uint32_t frameSize = math::CT::volume<
+                typename T_FrameType::SuperCellSize
+            >::type::value;
+            uint32_t numParLastFrame = numParticles % frameSize;
+            // numParLastFrame == 0 means the last frame is fully filled
```

Comment:
Maybe instead of the comment create an explanatory variable `bool isLastFrameFull = ( numParLastFrame == 0 );` ?

Reply:
I can do it but I need to keep the comment because it is not clear why zero means the frame is full. I will try to improve the comment

Reply:
As discussed offline, there is a bug when `numParticles == 0` - the function now returns frameSize and should actually return 0. I guess one could use smth like `return numParticles ? (1 + (numParticles - 1) % frameSize) : 0`;

Reply:
note: it must be `numParticles ? (1 + (numParticles - 1) % frameSize) : 0;` because `%` has less precedence than `+`.

Reply:
Does it? Cppreference states [the remainder is the same priority as multiplication](https://en.cppreference.com/w/cpp/language/operator_precedence).

Reply:
you are right, I looked to [this site](https://en.wikipedia.org/wiki/Operators_in_C_and_C%2B%2B) and missed that the `+` there is a unary add. OK I will fix it